crates/iota-indexer/src/optimistic_indexing.rs

```diff
@@ -0,0 +1,393 @@
+            {
+                let full_tx_data = CheckpointTransaction {
+                    transaction,
+                    effects,
+                    events,
+                    input_objects,
+                    output_objects,
+                };
+                self.index_transaction(&full_tx_data).await?;
+            }
+            _ => {
+                tracing::warn!(
+                    "Cannot optimistically index because of missing in/out objs for tx: {tx_digest}"
+                );
+            }
+        }
+
+        let tx_block_response = self
+            .wait_for_local_indexing(tx_digest, options.clone())
+            .await?;
+
+        Ok(IotaTransactionBlockResponseWithOptions {
+            response: tx_block_response,
+            options: options.unwrap_or_default(),
+        }
+        .into())
+    }
+
+    async fn wait_for_local_indexing(
+        &self,
+        tx_digest: TransactionDigest,
+        options: Option<IotaTransactionBlockResponseOptions>,
+    ) -> Result<IotaTransactionBlockResponse, IndexerError> {
+        let backoff = backoff::ExponentialBackoff {
+            max_elapsed_time: Some(Duration::from_secs(30)),
+            ..Default::default()
+        };
+
+        backoff::future::retry(backoff, async || {
+            let tx_block_response = self
+                .indexer_reader
+                .multi_get_transaction_block_response_in_blocking_task(
+                    vec![tx_digest],
+                    options.clone().unwrap_or_default(),
+                )
+                .await
+                .map_err(|e| backoff::Error::Transient {
+                    err: e,
+                    retry_after: None,
+                })?
+                .pop();
+
+            match tx_block_response {
+                Some(tx_block_response) => Ok(tx_block_response),
+                None => Err(backoff::Error::Transient {
+                    err: IndexerError::PostgresRead("Transaction not present in DB".to_string()),
+                    retry_after: None,
+                }),
+            }
+        })
+        .await
+    }
+
+    async fn index_transaction(
+        &self,
+        full_tx_data: &CheckpointTransaction,
+    ) -> Result<(), IndexerError> {
+        let assigned_global_order = self
+            .assign_optimistic_tx_global_order(full_tx_data.transaction.digest())
+            .await?;
+
+        let Some(assigned_global_order) = assigned_global_order else {
+            // Global order was assigned earlier by other indexing process, we avoid double
+            // or concurrent indexing and return
+            return Ok(());
+        };
+
+        let extractor = TransactionExtractor::new(
+            full_tx_data,
+            assigned_global_order
+                .optimistic_sequence_number
+                .expect("Optimistic sequence number is always set for data read from DB")
+                .try_into()
+                .unwrap(),
```

Comment:
We should return an internal-server error instead.

Reply:
Actually, why `optimistic_sequence_number` in TxGlobalOrder is i64? We could turn it into u64 and avoid a conversion here.